### PR TITLE
Read/write nested data as flattened parameter name to and from toml file

### PR DESCRIPTION
**Pull request by sverhoeven**
Refs #47

The haddock3 catalogs have not been updated. The different ways of supported flattening described at [docs/tomlSchema.md](https://github.com/i-VRESSE/workflow-builder/blob/write-expandable2toml/docs/tomlSchema.md) can be tested by loading the kitchensink catalog and configuring [node4](https://github.com/i-VRESSE/workflow-builder/blob/effed486e051c32da1324d484d3d2b7135df50f6/public/catalog/kitchensink.yaml#L109-L212).

Example worklow
[workflow (16).zip](https://github.com/i-VRESSE/workflow-builder/files/8215779/workflow.16.zip)

### CODE REVIEW

**Review by Peter9192 — COMMENTED** [review]
Haven't looked at the code yet, but from playing with the preview I noticed a couple of interesting things. 

## Point 1
With nested indexing the layout is a bit confusing. I click the first blue +, which effectively adds an empty nested array. As a user, the only visual feedback I get is the appearance of another blue + and a minus.

image

Since adding an empty array doesn't make much sense, does it make sense to automatically also create the first item in the nested array (and perhaps fill it with defaults or so, see #54)?

## Point 2

The behavior when leaving boxes empty is also a bit tricky:

image

Note that 1_1 is skipped in the toml file, since it is apparently a default value, but the fact that the count then starts at 1_2 suggests that 1_1 is still significant. But what happens with the last one? It's also pruned from the toml file, but it can never be known that there would have been a 2_2 as well.

File: src/toml.ts [hunk 1]
```diff
@@ -1,65 +1,175 @@
...
+function parameters2toml (parameters: IParameters, tomlSchema: TomlObjectSchema): any {
+  const tomledParameters: Record<string, unknown> = {}
+  Object.entries(parameters).forEach(([k, v]) => {
+    const isArray = Array.isArray(v)
+    const hasTomlSchema = k in tomlSchema
+    const isIndexed = hasTomlSchema && tomlSchema[k].indexed === true
+    const isSectioned = hasTomlSchema && tomlSchema[k].sectioned === true
+    const isItemsSectioned = hasTomlSchema && tomlSchema[k].items?.sectioned === true
+    const hasNestedTomlSchema = hasTomlSchema && isObject(tomlSchema[k].items?.properties)
+    const nestedTomlSchema = hasNestedTomlSchema ? tomlSchema[k].items?.properties : {}
```
Comment (Peter9192) [inline src/toml.ts:31]: I like these names, they really help to make the code readable.

File: src/toml.test.ts [hunk 2]
```diff
@@ -115,4 +367,421 @@ foo = 'fizz'
     }
     expect(result).toEqual(expected)
   })
+
+  it('should read complex values by default', () => {
```
Comment (Peter9192) [inline src/toml.test.ts:371]: Would it make sense to combine the tests for back and forth conversion, to save some duplicaiton?

**Review by Peter9192 — APPROVED** [review]
Looks very good, apart from my previous concern about the trailing empty array items I think it is good to go.

File: docs/tomlSchema.md [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+In toml
+
+```toml
+[[name]]
+something = 11
+else = 22
+
+[[name]]
+something = 33
+else = 44
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply docs/tomlSchema.md:142]: Using `[[name]]` is part of the toml spec and is called a [Array of tables](https://toml.io/en/v1.0.0#array-of-tables). So any toml library will know the written order is also the array order.

File: src/toml.ts (hunk 1 above)
Comment (sverhoeven) [reply src/toml.ts:31]: thanks

File: src/toml.test.ts (hunk 2 above)
Comment (sverhoeven) [reply src/toml.test.ts:371]: I would like to keep each test easily debug-able and make it do one thing. Having variables defined centrally or somehow combined tests would make it harder to see what a test does.

File: docs/tomlSchema.md [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+Below are examples of supported toml schema snippets.
+
+## Array of scalars
+
+In toml
+
+```toml
+param_1 = 11
+param_2 = 22
+param_3 = 33
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply docs/tomlSchema.md:58]: I chose those number to easily map the indices to the values. Using arbitrary numbers would make it harder to investigate any errors.

File: docs/tomlSchema.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (Peter9192) [reply docs/tomlSchema.md:142]: Thanks for the link

File: docs/tomlSchema.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (Peter9192) [reply docs/tomlSchema.md:58]: But this is the documentation, not testing code or so

**Comment by sverhoeven** [comment]
Thanks for the review. 

For point 1 the table widget in #53 and 2d matrix widget https://github.com/i-VRESSE/workflow-builder/issues/56 will help.

For point 2 trailing items which are equal to defaults and absent in toml is something the app running the toml needs to deal with.

**Pull request merged**